lib/msf/ui/console/command_dispatcher/exploit.rb

```diff
@@ -179,6 +114,84 @@ def cmd_exploit_tabs(str, words)
     tab_complete_generic(fmt, str, words)
   end
 
+  #
+  # Launches an exploitation attempt.
+  #
+  def cmd_exploit(*args)
+    force = false
+    opts = {
+      'Encoder'     => mod.datastore['ENCODER'],
+      'Payload'     => mod.datastore['PAYLOAD'],
+      'Target'      => mod.datastore['TARGET'],
+      'Nop'         => mod.datastore['NOP'],
+      'OptionStr'   => nil,
+      'LocalInput'  => driver.input,
+      'LocalOutput' => driver.output,
+      'RunAsJob'    => false,
+      'Background'  => false,
+      'Force'       => false
+    }
+
+    if mod.passive?
+      opts['RunAsJob'] = true
+    end
+
+    @@exploit_opts.parse(args) { |opt, idx, val|
+      case opt
+        when '-e'
+          opts['Encoder'] = val
+        when '-f'
+          force = true
+        when '-j'
+          opts['RunAsJob'] = true
+        when '-J'
+          opts['RunAsJob'] = false
+        when '-n'
+          opts['Nop'] = val
+        when '-o'
+          opts['OpttionStr'] = val
+        when '-p'
+          opts['Payload'] = val
+        when '-t'
+          opts['Target'] = val.to_i
+        when '-z'
+          opts['Background'] = true
+        when '-h'
+          cmd_exploit_help
+          return false
+      end
+    }
+
+    minrank = RankingName.invert[framework.datastore['MinimumRank']] || 0
+    if minrank > mod.rank
+      if force
+        print_status("Forcing #{mod.refname} to run despite MinimumRank '#{framework.datastore['MinimumRank']}'")
+        ilog("Forcing #{mod.refname} to run despite MinimumRank '#{framework.datastore['MinimumRank']}'", 'core')
+      else
+        print_error("This exploit is below the minimum rank, '#{framework.datastore['MinimumRank']}'.")
+        print_error("If you really want to run it, do 'exploit -f' or")
+        print_error("setg MinimumRank to something lower ('manual' is")
+        print_error("the lowest and would allow running all exploits).")
+        return
+      end
+    end
+
+    if !opts['Payload']
+      opts['Payload'] = Exploit.choose_payload(mod, opts['Target'])
+    end
+
+    rhosts = mod.datastore['RHOSTS']
+    if rhosts
+      Rex::Socket::RangeWalker.new(rhosts).each do |rhost|
+        nmod = mod.replicant
```

Comment:
the way this patch works and avoids clobbering the global RHOSTS value when it sets RHOSTS per-run to a single value is to call 'replicant', which makes a fully copy. This is opposed to aux scanner which assumes that RHOST and RHOSTS coexist. But to hide that sausage-making in the background, we have hundreds of unregister RHOST calls in aux scanners, that we could remove after this.